File: java/src/test/java/challenges/aoc2024/day4/model/WordsearchTest.java
package challenges.aoc2024.day4.model;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.List;

import org.junit.jupiter.api.Test;

import model.CoordinatePair;

class WordsearchTest {
    
    @Test
    void shouldBuildWordsearchObjectFromRows() {
        // Given
        final List<String> rows = List.of("HELLO", "THERE");

        // When
        final Wordsearch wordsearch = Wordsearch.of(rows);

        // Then
        assertEquals(ImmutableWordsearch.class, wordsearch.getClass());
        assertEquals(10, wordsearch.getLetters().size());
        
        final WordsearchLetter letterH = wordsearch.getLetters().get(0);
        assertEquals("H", letterH.getLetter());
        assertEquals(CoordinatePair.of(0, 0), letterH.getPostion());

        final WordsearchLetter lastE = wordsearch.getLetters().get(9);
        assertEquals("E", lastE.getLetter());
        assertEquals(CoordinatePair.of(4, 1), lastE.getPostion());
    }
}

File: scala/src/test/scala/challenges/aoc2024/day4/model/WordsearchTest.scala
package challenges.aoc2024.day4.model

import model.CoordinatePair
import org.scalatest.funsuite.AnyFunSuite

class WordsearchTest extends AnyFunSuite {

  test("should build Wordsearch object from rows") {
    // Given
    val rows: List[String] = List("HELLO", "THERE")

    // When
    val wordsearch: Wordsearch = Wordsearch.of(rows)

    // Then
    assert(wordsearch.isInstanceOf[Wordsearch])
    assert(10 == wordsearch.letters.size)

    val letterH: WordsearchLetter = wordsearch.letters.head
    assert("H" == letterH.letter)
    assert(CoordinatePair(0, 0) == letterH.position)

    val letterE: WordsearchLetter = wordsearch.letters(9)
    assert("E" == letterE.letter)
    assert(CoordinatePair(4, 1) == letterE.position)
  }
}
